### PR TITLE
[ACNA-1229] utils.getActionUrls returns different action name format if action is in the first package

### LINKED ISSUE #64: utils.getActionUrls returns different action name format if action is in the first package
```
packages:
  pkg1:
     actions:
       my-action:
         web: true
  pkg2:
     actions:
       my-action2:
         web: true
```

would return url object like:
```
{
  'myaction': <url>,
  'pkg2/my-action2': <url>
}
```
the expected should be like:
```
{
  'pkg1/myaction': <url>,
  'pkg2/my-action2': <url>
}
```

Why does this matter ?
Commands calling this method shouldn't need to treat action names differently depending on the package position in the manifest

### PR DESCRIPTION
**Pull request by florind12**
## Description

The dist output has changed, each action is now under its specific package. Updated the build & deploy function to support the change
Note: In order to support legacy code, the output of dist will include the old path as well. Further, the old path will not be processed in the deployment process. 
Actions and Sequences keys for generated URLs should now include the package's name as well. 

## Related Issue
MERGE ONLY WITH:
https://github.com/adobe/aio-cli-plugin-app/pull/495 
https://github.com/adobe/generator-aio-app/pull/178

 

Fixes #64

## Motivation and Context
Commands calling this method shouldn't need to treat action names differently depending on the package position in the manifest

## How Has This Been Tested?
Unit testing/Manual testing. 

Prerequisites for manual testing:
https://github.com/adobe/aio-cli-plugin-app/pull/495 
https://github.com/adobe/generator-aio-app/pull/178

- Generate a new `aio` project. 
- `aio app build` should build actions under:
   1. the new path: `dist/{packageName}/{actionBuild}`
   2. old path (default packageName): `dist/{actionBuild}` 
   
- given only one `generic` action is configured, `aio app deploy` should deploy only one action. 
- given an `asset compute` boostraped project,  `aio app test` should run succesfully. 
- given an `experience cloud` boostraped project,  `aio app run` should run succesfully & 
- `./src/dx-excshell-1/web-src/src/config.json` file should contain url keys in `pkgName/actionName` format.
For default package actions, url keys are in `actionaName` format for legacy support. 

## Screenshots (if appropriate):
image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by shazron — DISMISSED** [review]

**Review by moritzraho — COMMENTED** [review]
DO NOT MERGE THIS NOW

Actually this is a bigger change than expected for three reasons: 
- It will require changes in app plugin https://github.com/adobe/aio-cli-plugin-app/search?q=getActionUrls
- This is a breaking change for aio-lib-runtime and not a bug fix
- This is also a breaking change in the app plugin as it will generate urls to `web-src/config.json` in a different format. We might also need changes in the generator's UI

**Comment by florind12** [comment]
New info: 
These changes will break `aio-cli-plugin-asset-compute`. 
To discuss with Nui team about the change.

**Comment by shazron** [comment]
Note: the breakage is in the output file paths in the `dist` folder (which is a build artifact and is not guaranteed to be a in a certain structure anyway)

**Comment by florind12** [comment]
> Note: the breakage is in the output file paths in the `dist` folder (which is a build artifact and is not guaranteed to be a in a certain structure anyway)

So I understand that we didn't include the package name for the first package action in order the keep compatibility with legacy code, at the point where multipackage support was introduced, right? 

Since then we can consider that there is code built around on the actual dist gets generated. So I propose we keep the old dist and drop the new changes regarding dist.

**Comment by shazron** [comment]
> > Note: the breakage is in the output file paths in the `dist` folder (which is a build artifact and is not guaranteed to be a in a certain structure anyway)
> 
> So I understand that we didn't include the package name for the first package action in order the keep compatibility with legacy code, at the point where multipackage support was introduced, right?
> 
> Since then we can consider that there is code built around on the actual `dist` which gets generated. So I propose we keep the old dist and drop the new changes regarding dist. This way we maintain the compatibility

We might be kicking technical debt down the road however -- if a dev moves packages around in the runtime manifest, there will be no semblance of predictability in the `dist` folder, since the "first" package changed. Then if someone depends on the old structure, their script will break anyway since it will grab the wrong action. Having all actions in their respective package folder removes all ambiguity. 

My proposal is for devs that are used to the old structure -- to check for the file at the old location, and if it does not exist, check for it at the new location. This would be temporary until everyone updates their code/plugin.

**Comment by shazron** [comment]
I think the only way to make everyone happy, like the way we do it for config.json (have the action without package, and with package name as the key name), is to output to both locations (old and new)? WDYT? There won't be a perf impact if we just copy the files. And we give the devs that rely on the old location time to migrate.

**Comment by florind12** [comment]
> I think the only way to make everyone happy, like the way we do it for config.json (have the action without package, and with package name as the key name for the first package), is to output to both locations (old and new)? WDYT? There won't be a perf impact if we just copy the files. And we give the devs that rely on the old location time to migrate.

We could do but it with some extra work on getting around this new copy package in the build and deploys phase.
As we want to zip this extra action but we don't want to deploy it, right?

**Comment by shazron** [comment]
> We could do it but with some extra work on getting around this new copy in the build and deploys phase. As we want to zip this extra action but we don't want to deploy it, right?

Yes, it's only for backwards compat.

**Comment by Himavanth** [comment]
This is good and a required change to remove ambiguity without adding tech debt. I think it is safest if we can release this in a CLI major. 
BTW would this break apps which are changing the build artifact in pre-deploy hook ?

**Comment by florind12** [comment]
> This is good and a required change to remove ambiguity without adding tech debt. I think it is safest if we can release this in a CLI major. BTW would this break apps which are changing the build artifact in pre-deploy hook ?

That's a very good observation. Considering that there are people that might alter the old artifact, a link between the old path and the new path should exist.

File: src/build-actions.js
Comment (sandeep-paliwal) [inline src/build-actions.js:237]: @florind12 why do we need symlinks here?

**Comment by sandeep-paliwal** [comment]
@florind12 There are some conflicts to PR which needs to be resolved.
Otherwise LGTM. Will approve once I try out the changes.

**Review by sandeep-paliwal — APPROVED** [review]
LGTM

File: src/build-actions.js
Comment (purplecabbage) [reply src/build-actions.js:238]: symlinks here were a mistake. they break windows as usually the tools do not have permission to create them.
[ACNA-1637](https://jira.corp.adobe.com/browse/ACNA-1637)

**Pull request merged**
